R/submit_ipr.R

```diff
@@ -1,145 +1,417 @@
   data <- list(
-    email=user_email,
-    title="test",
-    goterms="false",
-    pathways="false",
-    stype="p",
-    appl="PfamA",
-    sequence=sequence_str
+    email = user_email,
+    title = "riprscan",
+    goterms = "false",
+    pathways = "false",
+    stype = "p",
+    appl = c("PfamA", "CDD"),
+    sequence = sequence_str
   )
-
+  ### global
+  total_jobs <<- total_jobs + 1
   res_post <- httr::POST(url_post, body = data, httr::add_headers(headers_post))
-  cat("Request: \n", sep = "")
+  job_id <- rawToChar(res_post$content)
+  cat("Request info for ", job_id, ": \n", sep = "")
   print(res_post$request)
-  if (as.integer(res_post$status) != 200L) {
-    message <- paste0("POST error with sequence: ", fasta_header, "\n",
-                      "Status code: ", as.integer(res_post$status), "\n")
+
+  status <- ifelse(as.integer(res_post$status) == 200L, "running", "failed")
+  if (status == "failed") {
+    message <- paste0(
+      "POST error with sequence: ", fasta_header, "\n",
+      "Status code: ", as.integer(res_post$status), "\n"
+    )
     warning(message)
-    return(NULL)
-  }
-  else {
-    cat("POST status: ", as.integer(res_post$status), "\n", sep = "")
+    .add_job_status_table(basename(path2seq), job_id, status)
+    return("failed")
+  } else {
+    .add_job_status_table(basename(path2seq), job_id, status)
+    return(job_id)
   }
-  job_id <- rawToChar(res_post$content)
-  cat("job_id: ", job_id, "\n", sep = "")
+}
+
 
-  ### GET
+.poll_job <- function(job_code) {
+  cat("Polling job: ", job_code, "\n", sep = "")
   url_get_base <- "https://www.ebi.ac.uk/Tools/services/rest/iprscan5/result/"
-  url_get <- paste0(url_get_base, job_id, "/tsv")
+  url_get <- paste0(url_get_base, job_code, "/tsv")
   header_get <- "Accept: text/tab-separated-values"
+  res_get <- httr::GET(url_get, httr::add_headers(header_get))
 
-  status <- -1L
-  n_polls <- 1L
-  # Try poll 30 times OR until status is complete
-  while (status != 200L && n_polls < 180L) {
-    res_get <- httr::GET(url_get, httr::add_headers(header_get))
-    status <- as.integer(res_get$status_code)
-    cat("GET poll #", n_polls, "\t", "Status: ", status, "\n", sep = "")
-    n_polls <- n_polls + 1L
-    Sys.sleep(5L)
-  }
-  if (status != 200L) {
-    message <- paste0("Failed to retrieve job results for <",
-                      job_id,
-                      ">",
-                      "\n",
-                      "Job exceeded 30 minutes runtime.")
-    warning(message)
+  status <- as.integer(res_get$status_code)
+  status <- ifelse(status == 200L, "completed", "running")
+
+  # update N polls & latest poll time
+  tb_status <<- tb_status %>%
+    dplyr::mutate(polls = ifelse(job_id == job_code, polls + 1, polls)) %>%
+    dplyr::mutate(t_latest_poll = ifelse(job_id == job_code, as.POSIXct(Sys.time()), t_latest_poll))
+
+  return(status)
+}
+
+.update_status_table <- function(job_code, status_update) {
+  # change status for a given job_id row
```

Comment:
`job_code` vs `job_id` -- add a little clarification for future functions that depend on this understanding!

Reply:
Yeah good point. I actually wasn't sure if R would bug/complain if I try to set a list name pointer called `job_id` to a value also called `job_id`. I just did a small test, and it seems like R is able to separate the two.

Example
```
construct_list <- function(job_id) {
  new_list <- list(
    job_id = job_id,
    status = status
  )
  return(new_list)
}
```
I now know that this will work as intended where the name stays as `job_id` while the value is set the to the `job_id` argument passed into the function.